services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/DocumentTableSchema.java

```diff
@@ -0,0 +1,220 @@
+import software.amazon.awssdk.enhanced.dynamodb.TableSchema;
+import software.amazon.awssdk.enhanced.dynamodb.document.EnhancedDocument;
+import software.amazon.awssdk.enhanced.dynamodb.internal.converter.ConverterProviderResolver;
+import software.amazon.awssdk.enhanced.dynamodb.internal.document.DefaultEnhancedDocument;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+
+
+/**
+ * Implementation of {@link TableSchema} that builds a table schema based on DynamoDB Items.
+ * <p>
+ * In Amazon DynamoDB, an item is a collection of attributes. Each attribute has a name and a value. An attribute value can be a
+ * scalar, a set, or a document type
+ * <p>
+ * A DocumentTableSchema is used to create a {@link DynamoDbTable} which provides read and writes access to DynamoDB table as
+ * {@link EnhancedDocument}.
+ * <p> DocumentTableSchema specifying primaryKey, sortKey and a customAttributeConverter can be created as below
+ * {@snippet :
+ * DocumentTableSchema documentTableSchema = DocumentTableSchema.builder()
+ * .primaryKey("sampleHashKey", AttributeValueType.S)
+ * .sortKey("sampleSortKey", AttributeValueType.S)
+ * .attributeConverterProviders(customAttributeConveter, AttributeConverterProvider.defaultProvider())
+ * .build();
+ *}
+ * <p> DocumentTableSchema can also be created without specifying primaryKey and sortKey in which cases the
+ * {@link TableMetadata} of DocumentTableSchema will error if we try to access attributes from metaData.
+ * Also if attributeConverterProviders are not provided then {@link DefaultAttributeConverterProvider} will be used
+ * {@snippet :
+ * DocumentTableSchema documentTableSchema = DocumentTableSchema.builder().build();
+ *}
+ *
+ * @see <a href="https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithItems.html" target="_top">Working
+ * with items and attributes</a>
+ */
+@SdkPublicApi
+public final class DocumentTableSchema implements TableSchema<EnhancedDocument> {
+
+    private final TableMetadata tableMetadata;
+    private final List<AttributeConverterProvider> attributeConverterProviders;
+
+    private DocumentTableSchema(Builder builder) {
+        this.attributeConverterProviders = builder.attributeConverterProviders;
+        this.tableMetadata = builder.staticTableMetaDataBuilder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public EnhancedDocument mapToItem(Map<String, AttributeValue> attributeMap) {
+        if (attributeMap == null) {
+            return null;
+        }
+        DefaultEnhancedDocument.DefaultBuilder builder = DefaultEnhancedDocument.builder();
+        return builder.attributeValueMap(attributeMap)
+                      .attributeConverterProviders(attributeConverterProviders)
+                      .build();
+    }
+
+    @Override
+    public Map<String, AttributeValue> itemToMap(EnhancedDocument item, boolean ignoreNulls) {
```

Comment:
yes , the "ignoreNulls" is used for Mapped Class/Beans , in case of Document the enhancedDocument returns Null if Attribute is absent or if attribute is explicitly set to null  (Same as Json Null property) .
And Unlike Mapped classes where the CLient is aware of complete Schema and marking the Attributes as Null makes sense , however in Document Schema the client is not aware of complete Schema and if attributes are not present it means they are null in case of Document Schema

Reply:
Gotcha, can we add some documentation around it? basically telling users setting this has no effect

Reply:
done